File: prob1.py
#  Time Complexity : O(n)
# Space Complexity :O(1)
# Leet Code: Yes

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        lp = 1
        n = len(nums)
        # initialize array
        arr = [None for _ in range(n)]
        arr[0] = 1

        # do left product
        for i in range(1, n):
            lp *= nums[i-1]
            arr[i] = lp

        #do right product and save the output in the arr
        rp = 1
        for i in range(n-2, -1, -1):
            rp *= nums[i+1]
            arr[i] *= rp

        return arr

File: prob1.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Compute the left products of each element and store them in the leftProducts array.
// Compute the right products of each element and store them in the rightProducts array.
// Multiply the corresponding left and right products for each element to get the final output array.

public int[] productExceptSelf(int[] nums) {
    int n = nums.length;
    int[] leftProducts = new int[n];
    int[] rightProducts = new int[n];
    int[] output = new int[n];
    
    leftProducts[0] = 1;
    for (int i = 1; i < n; i++) {
        leftProducts[i] = leftProducts[i - 1] * nums[i - 1];
    }
    
    rightProducts[n - 1] = 1;
    for (int i = n - 2; i >= 0; i--) {
        rightProducts[i] = rightProducts[i + 1] * nums[i + 1];
    }
    
    for (int i = 0; i < n; i++) {
        output[i] = leftProducts[i] * rightProducts[i];
    }
    
    return output;
}

